### PR TITLE
Add Whole Word Masking and Padding Strategy to DataCollatorForLanguageModeling

### LINKED ISSUE #12998: DataCollatorForWholeWordMask does not return attention_mask
Hi,

** DataCollatorForWholeWordMask** does not output `attention_mask`. According to the `__call__` method:
`return {"input_ids": inputs, "labels": labels}`. 

Is there a peculiar motivation behind it or is a small bug? From where I see, when we do the pre-training, most instances will **not** be of the same length and applying _Attention_  for all the tokens (including the padding) may cause imprecise results.

### PR DESCRIPTION
**Pull request by rjgleaton**
Sorry about the previous PR. I made a git error when trying to amend my commits.

# What does this PR do?

Main contribution is adding Whole Word Masking to DataCollatorForLanguageModeling. This is to fix the issues mentioned in #12998 where the previous implementation only returned input_ids and labels. This implementation should pass through all values from the input mappings. Because I feel it's more intuitive and easier to maintain I've added an arg to DataCollatorForLanguageModeling that allows for whole word masking. The DataCollatorForWholeWordMasking collator has been largely deprecated and is now just a rapper for the language modeling collator where mlm and whole_word_mask are forced to true.

This implementation depends on offset mappings for determining if a given pair of tokens belong to the same word. As such, it only natively works for FastTokenizers. I'm not sure there's a tokenizer agnostic way to do this with non-fast tokenizers atm. 

~~Additionally I've adding an argument for specifying PaddingStrategy to DataCollatorForLanguageModeling. This is a minor QOL improvement that makes it more consistent with other collators.~~

Fixes #12998

## Before submitting
- [x] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

@sgugger, @Rocketknight1

### CODE REVIEW

File: src/transformers/data/data_collator.py [hunk 1]
```diff
@@ -769,10 +850,24 @@ class DataCollatorForLanguageModeling(DataCollatorMixin):
...
+        padding (`bool`, `str` or [`~utils.PaddingStrategy`], *optional*, defaults to `True`):
+            Select a strategy to pad the returned sequences according to the model's padding side and padding index
+            among:
+
+            - `True` or `'longest'` (default): Pad to the longest sequence in the batch (or no padding if only a single sequence
+              is provided).
+            - `'max_length'`: Pad to a maximum length specified with the argument `max_length` or to the maximum
+              acceptable input length for the model if that argument is not provided.
+            - `False` or `'do_not_pad'`: No padding (i.e., can output a batch with sequences of different
+              lengths).
```
Comment (Rocketknight1) [inline src/transformers/data/data_collator.py]: This is probably a nice QOL improvement, but since I don't think there are dependencies between padding and WWM, I think they should be two separate PRs! It's a lot easier to review if we keep WWM in a single dedicated PR, and then we can add a `padding` arg afterwards.

File: src/transformers/data/data_collator.py [hunk 2]
```diff
@@ -1176,371 +1324,112 @@ def numpy_mask_tokens(self, inputs: Any, special_tokens_mask: Optional[Any] = No
...
-            self.tokenizer.get_special_tokens_mask(val, already_has_special_tokens=True) for val in labels.tolist()
-        ]
-        probability_matrix.masked_fill_(torch.tensor(special_tokens_mask, dtype=torch.bool), value=0.0)
-        if self.tokenizer.pad_token is not None:
-            padding_mask = labels.eq(self.tokenizer.pad_token_id)
-            probability_matrix.masked_fill_(padding_mask, value=0.0)
+            for token_idx, (token_start, token_end) in enumerate(offset):
+                if special_tokens_mask[seq_idx][token_idx] == 1:
+                    word_ids.append(-1)
+                    prev_token_end = None
```
Comment (Rocketknight1) [inline src/transformers/data/data_collator.py]: This logic seems correct but a Python loop over the entire token array could be quite slow for longer sequences and big batches! Maybe we can vectorize somehow, for example by getting arrays of token starts/ends instead, and then:

```python
word_boundaries = token_ends[:-1] == token_starts[1:]
```

Of course, if we need Python list comprehensions to build the arrays in the first place, then the speed improvement might not be that significant.

**Review by Rocketknight1 — COMMENTED** [review]
Hi @rjgleaton, thank you for the PR! We did some discussion internally and we think having a WWM masker / collator that works across lots of tokenizer classes would be great to have in the library.

I made a couple of comments, mainly about splitting this into two PRs (one for WWM, the other for adding `PaddingStrategy`) since the diff is already quite hard to review. Excited to get this in the library, though!

File: src/transformers/data/data_collator.py (hunk 1 above)
Comment (rjgleaton) [reply src/transformers/data/data_collator.py]: That's fair - I've taken it out of this PR. I'll make a quick issue to add it and fix the issue in a second PR.

File: src/transformers/data/data_collator.py (hunk 2 above)
Comment (rjgleaton) [reply src/transformers/data/data_collator.py]: I've updated both `_calc_word_ids_and_prob_mask` and `_whole_word_mask` to use numpy arrays instead of lists so the operations can be vectorized. 

For some reason I previously thought numpy wasn't an _explicit_ dependency, but that's obviously incorrect.

**Comment by rjgleaton** [comment]
This is just a quick callout - I added some unit tests for the two new functions I added since the logic isn't _super_ trivial. I'm not entirely clear on what conventions y'all prefer for creating unit tests, so please lmk if they need to be updated.

**Comment by rjgleaton** [comment]
@Rocketknight1 pinging about this PR again. No rush, just making sure it doesn't get lost.

File: src/transformers/data/data_collator.py [hunk 3]
```diff
@@ -1330,6 +1322,17 @@ def tolist(x) -> list[Any]:
...
 
+def to_numpy(x) -> np.ndarray[Any]:
+    if isinstance(x, np.ndarray):
+        return x
+    elif hasattr(x, "numpy"):
+        return x.numpy()
+    elif hasattr(x, "detach"):
+        return x.detach().cpu().numpy()
+    else:
+        return np.array(x)
```
Comment (Rocketknight1) [inline src/transformers/data/data_collator.py]: I'm a little unsure about this function! Torch tensors that require grad or are on the CUDA device will fail if you call `tensor.numpy()`. I think the `elif hasattr(x, "detach")` branch will never be reached because all torch tensors that have `detach()` also have `numpy()`, so those tensors will just try to call `numpy()` and then throw an error.

I realize you're trying to write something that works for TF tensors as well, but our support for TF and JAX is now deprecated, and I prefer not to touch that code too much because we no longer have tests guaranteeing correctness. Maybe we can get rid of this and not try to support WWM on TF/JAX?

**Review by Rocketknight1 — APPROVED** [review]
Yes, I'm sorry for losing track of this! The GPT-OSS launch made me drop everything else for a week or two, lol. The new changes look good, but I made one comment. Sorry to keep iterating like this!

File: src/transformers/data/data_collator.py (hunk 3 above)
Comment (rjgleaton) [reply src/transformers/data/data_collator.py]: Should be fixed now - I removed the whole word masking code from tf_call, and added a value error if that function is called with whole_word_masking set to true since whole word masking can modify the init.

**Comment by Rocketknight1** [comment]
Yes, I think it looks good now! There's one final thing - we have a big PR to finally remove almost all of the TF and Jax code in the codebase here: #40760

Because of how large it is, we're trying not to merge other PRs that conflict with it right now. Can you wait until #40760 is merged, then rebase and resolve any conflicts here, then finally ping me and we can merge this? I'm sorry for all the delays, but we're almost there!

**Comment by Rocketknight1** [comment]
#40760 is merged! You can rebase this PR and ping me for a final review

**Comment by rjgleaton** [comment]
> #40760 is merged! You can rebase this PR and ping me for a final review

@Rocketknight1 Should be good to review again. Some tests seem to be failing, though I don't _think_ they're related to my dev. Let me know if that's unexpected.

**Review by Rocketknight1 — APPROVED** [review]
Yes, this looks ready to merge now! Removing TF also cleaned up the diff a lot. Thank you for the nice PR, and for your patience with the delays and rebases!
